src/sentry/mail/notifications.py

```diff
@@ -72,24 +85,27 @@ def get_context(
 
 @register_notification_provider(ExternalProviders.EMAIL)
 def send_notification_as_email(
-    notification: ActivityNotification,
-    users: Mapping[User, int],
+    notification: BaseNotification,
+    users: Set[User],
     shared_context: Mapping[str, Any],
+    extra_context_by_user_id: Optional[Mapping[int, Mapping[str, Any]]],
 ) -> None:
     headers = get_headers(notification)
     subject = get_subject_with_prefix(notification)
     type = get_email_type(notification)
 
-    for user, reason in users.items():
+    for user in users:
+        extra_context = (extra_context_by_user_id or {}).get(user.id, {})
+        log_message(notification, user)
         msg = MessageBuilder(
             subject=subject,
+            context=get_context(notification, user, shared_context, extra_context),
             template=notification.get_template(),
             html_template=notification.get_html_template(),
             headers=headers,
-            type=type,
-            context=get_context(notification, user, reason, shared_context),
-            reference=notification.activity,
+            reference=notification.get_reference(),
```

Comment:
why is `notification.activity` being replaced with `self.group`?

Reply:
oops that's a bug!

Reply:
I've fixed it but it was passing tests because `reply_reference` was overwriting `reference`.